flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java

```diff
@@ -71,10 +70,26 @@ public void cleanCheckpoint(
             boolean shouldDiscard,
             Runnable postCleanAction,
             Executor executor) {
-        Checkpoint.DiscardObject discardObject =
-                shouldDiscard ? checkpoint.markAsDiscarded() : Checkpoint.NOOP_DISCARD_OBJECT;
-
-        cleanup(checkpoint, discardObject::discard, postCleanAction, executor);
+        if (shouldDiscard) {
+            incrementNumberOfCheckpointsToClean();
+            checkpoint
+                    .markAsDiscarded()
+                    .discardAsync(executor)
+                    .handle(
+                            (Object outerIgnored, Throwable outerThrowable) -> {
+                                if (outerThrowable != null) {
+                                    LOG.warn(
+                                            "Could not properly discard completed checkpoint {}.",
+                                            checkpoint.getCheckpointID(),
+                                            outerThrowable);
+                                }
+                                decrementNumberOfCheckpointsToClean();
```

Comment:
`decrementNumberOfCheckpointsToClean` this should be also called if `shouldDiscard == false`, right?

Reply:
You have just said it yourself. Old code for `shouldDiscard=false` is doing:
```
try {
    cleanupAction.run();
} catch (...) {
    ...
} finally {
    decrementNumberOfCheckpointsToClean();
    postCleanupAction.run();
}
```
with no exception and the `cleanupAction` being `NOOP`, code simplifies to:
```
    decrementNumberOfCheckpointsToClean();
    postCleanupAction.run();
```

`decrementNumberOfCheckpointsToClean` is clearly there.

Reply:
i think the confusion is from that I moved the incrementNumberOfCheckpointsToClean() to only shouldDiscard=true. so when shouldDiscard=false, there is no increment, thus no decrement as well. Or I am seriously missing something?

Reply:
Ahhh, yes you are right. Thanks for the explanation :)